### PR TITLE
Sync with upstream `v1.29.4`

### LINKED ISSUE #326: [rel-v1.29] Sync with Upstream `1.29.4`
**What would you like to be added**:
Sync the branch with upstream `1.29.4`.

**Why is this needed**:
Required to incorporate https://github.com/kubernetes/autoscaler/pull/6757. We have seen problems caused due to the shadowed declaration in live issue 5587

### PR DESCRIPTION
**Pull request by rishabh-11**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
Fixes #326 

**Special notes for your reviewer**:
IT pass for AWS

**Release note**:

```other operator
Synced with upstream v1.29.4
```

### CODE REVIEW

File: cluster-autoscaler/cloudprovider/mcm/mcm_manager.go [hunk 1]
```diff
@@ -441,6 +441,7 @@ func (m *McmManager) SetMachineDeploymentSize(ctx context.Context, machinedeploy
 // DeleteMachines deletes the Machines and also reduces the desired replicas of the MachineDeployment in parallel.
 func (m *McmManager) DeleteMachines(targetMachineRefs []*Ref) error {
 	if len(targetMachineRefs) == 0 {
+		klog.V(2).Infof("[DeleteMachines] No machines to delete")
```
Comment (aaronfern) [inline cluster-autoscaler/cloudprovider/mcm/mcm_manager.go:444]: `Info` should be sufficient, no?
```suggestion
		klog.V(2).Info("[DeleteMachines] No machines to delete")
```

File: cluster-autoscaler/main.go [hunk 2]
```diff
@@ -608,14 +608,6 @@ func main() {
...
 	klog.V(1).Infof("Cluster Autoscaler %s", version.ClusterAutoscalerVersion)
 
-	// FORK-CHANGE: logging version of g/autoscaler as well
-	gardenerversion, err := os.ReadFile("VERSION")
-	if err != nil {
-		klog.Warningf("Error reading gardener autoscaler version, err: %s", err)
-	} else {
-		klog.V(1).Infof("Gardener Cluster Autoscaler %s", gardenerversion)
-	}
-
```
Comment (aaronfern) [inline cluster-autoscaler/main.go]: This is removed by error ig. Can you please re add this?

**Review by aaronfern — CHANGES_REQUESTED** [review]
Looks good in general

In addition to the two changes asked, I have one question, and one more change
1. Do we not update `SYNC-CHANGES` with patch versions?
2. Can you please update the docstring of [`machine_cloud_provider.GetOptions`](https://github.com/gardener/autoscaler/blob/machine-controller-manager-provider/cluster-autoscaler/cloudprovider/mcm/mcm_cloud_provider.go#L477-L479) to the new docstring [here](https://github.com/gardener/autoscaler/pull/327/files#diff-df1da4cc9174c37604da59b5db78a797e3b6450c71490f26a5b4a3f74263be12R235-R237)

File: cluster-autoscaler/main.go (hunk 2 above)
Comment (rishabh-11) [reply cluster-autoscaler/main.go]: Did not remove it by mistake. There is no `Version` file anymore. So we can either add a version file and then reintroduce this or we can get rid of this.

**Comment by rishabh-11** [comment]
> Do we not update SYNC-CHANGES with patch versions?

It is not needed IMO as the sync changes are for minor and not patch versions.

> Can you please update the docstring of [machine_cloud_provider.GetOptions](https://github.com/gardener/autoscaler/blob/machine-controller-manager-provider/cluster-autoscaler/cloudprovider/mcm/mcm_cloud_provider.go#L477-L479) to the new docstring [here](https://github.com/gardener/autoscaler/pull/327/files#diff-df1da4cc9174c37604da59b5db78a797e3b6450c71490f26a5b4a3f74263be12R235-R237)

The docstring is part of 1.30 and is also not present upstream. I think it is fine for us to keep it like this.

File: cluster-autoscaler/main.go (hunk 2 above)
Comment (rishabh-11) [reply cluster-autoscaler/main.go]: I checked and it is working fine. It uses the VERSION file at the root of the repo, not the version dir. I have added this piece of code back

**Comment by aaronfern** [comment]
> I think it is fine for us to keep it like this

Okay, fair enough for me then

**Review by aaronfern — APPROVED** [review]
Thanks for the changes
/lgtm

**Pull request merged**